### PR TITLE
Add COE expiry filter

### PR DESCRIPTION
**Pull request by nighoggDatatype**
Adds COE Expiry filter and test code based on convention from Customer attributes and other commands, where appropriate

### CODE REVIEW

File: src/main/java/seedu/address/logic/filters/CoeExpiryFilter.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    private boolean isCoeExpiryExpiringSoon(CoeExpiry coeExpiry) {
+        return coeExpiryThreshold.isAfter(coeExpiry.toDate());
+    }
+
+    @Override
+    public boolean test(Customer customer) {
+        requireNonNull(customer);
+        return customer.getCarsOwned()
+                       .values()
+                       .parallelStream()
```
Comment (rajobasu) [inline src/main/java/seedu/address/logic/filters/CoeExpiryFilter.java:64]: change this to normal stream, parallel stream just makes things slower on such a small scale.

File: src/main/java/seedu/address/logic/filters/CoeExpiryFilter.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+     *
+     * A special keyword "exp" is accepted as an alias for COE
+     * that expires in zero years or less (ie: Already expired).
+     *
+     * Note that {@code filterString} must either be "exp" or a non negative number.
+     *
+     * @param filterString The number of years that a COE can be valid for before
+     *                     being skipped.
+     */
+    public CoeExpiryFilter(String filterString) {
```
Comment (rajobasu) [inline src/main/java/seedu/address/logic/filters/CoeExpiryFilter.java:29]: good idea would be to trim the filterString that you get

File: src/main/java/seedu/address/model/customer/Car.java [hunk 3]
```diff
@@ -66,10 +66,7 @@ public String toString() {
 
     @Override
     public boolean equals(Object other) {
-        return other == this // short circuit if same object
-                || (other instanceof Car // instanceof handles nulls
-                && carBrand.equals(((Car) other).carBrand)
-                && carType.equals(((Car) other).carType)); // state check
+        return other == this; // Not checking carBrand or CarType since it shouldn't be unique
```
Comment (rajobasu) [inline src/main/java/seedu/address/model/customer/Car.java:69]: um, i dont get why this is so? why just check objects for equality, same information should have same object. I think you should add another field like numberplate if you are worried about collisions. What you can do is if there is no explicit numberplate then just assign a random number to the numberplate (best random number is just the current timestamp in nanoseconds).

File: src/test/java/seedu/address/testutil/CustomerBuilder.java [hunk 4]
```diff
@@ -109,16 +109,12 @@ public CustomerBuilder withDateOfBirth(String dateOfBirth) {
...
     /**
-     * Parses the {@code carsOwned} into a {@code Map<Car, CoeExpiry>} and set it to the {@code Customer} that we are
-     * building.
-     *
-     * @param carsOwned
-     * @return
+     * Adds a new pairing of {@code Car} and {@code CoeExpiry} to the
+     * {@code Customer} that we are building.
      */
-    public CustomerBuilder withCarsOwned(Map<Car, CoeExpiry> carsOwned) {
```
Comment (rajobasu) [inline src/test/java/seedu/address/testutil/CustomerBuilder.java:120]: why did you delete this? we want to support multiple cars owned by the customer, all the code is written like that.

**Review by rajobasu — CHANGES_REQUESTED** [review]

**Review by rajobasu — APPROVED** [review]
LGTM, might need to change some things later on though.

File: src/main/java/seedu/address/logic/filters/CoeExpiryFilter.java (hunk 1 above)
Comment (nighoggDatatype) [reply src/main/java/seedu/address/logic/filters/CoeExpiryFilter.java:64]: Will do later

File: src/main/java/seedu/address/logic/filters/CoeExpiryFilter.java (hunk 2 above)
Comment (nighoggDatatype) [reply src/main/java/seedu/address/logic/filters/CoeExpiryFilter.java:29]: Will put the trimming function in the ParseUtil class.

File: src/main/java/seedu/address/model/customer/Car.java (hunk 3 above)
Comment (OhJunMing) [reply src/main/java/seedu/address/model/customer/Car.java:69]: We will shelve number plate away first, its not necessary
